### PR TITLE
[formal-snapshots] skiplist checkpoint sync and verification

### PR DESCRIPTION
**Pull request by williampsmith**
## Description 

Checkpoint summary sync and verification now dominates the critical path for formal snapshot restore. Mysticeti is magnifying this due to the increased checkpointing rate. 

This PR aims to fix this by removing the requirement that we download and verify all checkpoints. 

If `--all-checkpoints` flag is provided, behavior is as it was before. If not, we simply download all end of epoch checkpoints, check signatures, and omit the step where we verify checkpoint chain continuity. The security argument for why this is ok is that the validators in the committee had to verify chain continuity, and a quorum of them signed against the given checkpoint digest. Signature verification confirms that the contents have not been altered, and the digest matching confirms that the chain was checked for consistency against this digest by the committee.

In general this will lead to a 2x speed increase (moreso when comparing post-mysticeti)

Note that if the flag is not provided, we use the old behavior, sacrificing longer restore times for more hardened security.

## Test plan 

Run with strict verification

```
target/release/sui-tool download-formal-snapshot --network mainnet --latest --path /opt/sui/db/authorities_db/full_node_db --genesis /opt/sui/mainnet/config/genesis.blob --no-sign-request --verify strict
Beginning formal snapshot restore to end of epoch 427, network: Mainnet, verification mode: Strict                                                                                        [00:06:42] ██████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 468 out of 468 .ref files done (ref files download complete)
[00:01:03] █████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 428/428 (Checkpoint summary sync is complete)
[00:00:00] █████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 428/428 (Checkpoint summary verification is complete)
[00:00:41] ██████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 468 out of 468 ref files checksummed (Checksumming complete)
[00:02:19] ████████████████████████████████████████████████████████████████████████████████████████████████████ 468 out of 468 ref files accumulated from snapshot (Accumulation complete)
[00:33:12] ████████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 468 out of 468 .obj files done (Objects download complete)
[00:00:00] ███████████████████████████████████████████████████████████████████████████████████████████████████ Verifying snapshot contents against root state hash (Verification complete)
[00:40:42] ███████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 266867624/266867624 (DB live object accumulation completed)DB live object state verification completed successfully!
Successfully restored state from snapshot at end of epoch 427
```

Run with full summary sync and verification (old default)
```
target/release/sui-tool download-formal-snapshot --network mainnet --latest --path /opt/sui/db/authorities_db/full_node_db --genesis /opt/sui/mainnet/config/genesis.blob --no-sign-request --all-checkpoints
Formal snapshot state verification completed successfully!
[00:00:41] ██████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 465 out of 465 ref files checksummed (Checksumming complete)
[00:02:23] ████████████████████████████████████████████████████████████████████████████████████████████████████ 465 out of 465 ref files accumulated from snapshot (Accumulation complete)
[00:29:15] ████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 465 out of 465 .obj files done
(Objects download complete)
[00:40:33] ████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 36549181/36549181(Checkpoint summary verification is complete)
```

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

**Comment by bmwill** [comment]
Would it be possible to retain a more paranoid mode which does the checkpoint chain verification?

**Comment by williampsmith** [comment]
@bmwill yes the paranoid is still the default. To use the less paranoid mode requires providing `--with-skiplist` explicitly. I'll make that more clear in the PR summary

File: crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs [hunk 1]
```diff
@@ -381,17 +381,28 @@ impl CheckpointExecutor {
...
+                    .expect("Failed to delete contents digest -> sequence number mapping");
+            } else {
+                // If this is directly after a snapshot restore with skiplisting,
+                // this is expected for the first `NUM_SAVED_FULL_CHECKPOINT_CONTENTS`
+                // checkpoints.
+                debug!(
+                    "Failed to fetch checkpoint with sequence number {:?}",
+                    prune_seq
+                );
+            }
```
Comment (williampsmith) [inline crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs:405]: cc @aschran @phoenix-o can wanted to double check the above with you to make sure you don't see an issue there.

**Comment by lxfind** [comment]
I think we should make this mode default? Most of the time people don't need the paranoid mode

**Comment by williampsmith** [comment]
> I think we should make this mode default? Most of the time people don't need the paranoid mode

Sure we can do that

File: crates/sui-tool/src/lib.rs [hunk 2]
```diff
@@ -726,32 +739,34 @@ fn start_summary_sync(
...
+                )
+                .await?;
+        } else {
+            archive_reader
+                .read_summaries_with_range(
+                    state_sync_store.clone(),
+                    s_start..last_checkpoint + 1,
+                    sync_checkpoint_counter,
+                    // rather than blocking on verify, sync all summaries first, then verify later
+                    false,
```
Comment (lxfind) [inline crates/sui-tool/src/lib.rs:757]: Is there still a case where true is passed? Trying to understand if you still need this argument

Comment (williampsmith) [reply crates/sui-tool/src/lib.rs:757]: No this is the only callsite, but I had thought to leave the option in `read_summaries_with_range` with verification for the sake of generality. I also get concerned that someone may use the fn in the future without realizing that it is leaving verification to the caller.

I'll remove the verify path in here, but be more explicit in the naming

File: crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs (hunk 1 above)
Comment (williampsmith) [reply crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs:405]: Chatted about this offline. Going to leave as is here.

**Review by lxfind — APPROVED** [review]

**Pull request merged**